Feature/pass through chunk to alignment call

rrahn opened the pull request:
Fixes #1372

The alignment executor will now call the execution handler directly with passing the chunk of indexed sequence pairs. This also means that now the return type is a vector which needs to be moved to the result buffer.

joergi-w commented on include/seqan3/alignment/pairwise/detail/concept.hpp:
```diff
@@ -14,35 +14,78 @@
+//!\endcond
+
+/*!\interface seqan3::detail::sequence_pair_range <>
+ * \brief A helper concept to check if a type is a range over seqan3::detail::sequence_pair's.
```
...over `seqan3::detail::sequence_pair`.

joergi-w commented on include/seqan3/alignment/pairwise/detail/concept.hpp:
```diff
@@ -14,35 +14,78 @@
+ *
+ * \details
+ *
+ * This concept checks if the given type models a std::ranges::forward_range over and that the value type of the
```
~over~

joergi-w commented on include/seqan3/alignment/pairwise/detail/concept.hpp:
```diff
@@ -14,35 +14,78 @@
     requires std::tuple_size_v<decltype(value)> == 2;
-    requires tuple_like<std::tuple_element_t<0, decltype(value)>>;
-    requires std::tuple_size_v<std::tuple_element_t<0, decltype(value)>> == 2;
+    requires sequence_pair<std::tuple_element_t<0, decltype(value)>>;
```
Idea: Should we require some numeric/integral type or at least EqualityComparable for the index? Otherwise users are able to set weird types, which may cause side effects at some point.

rrahn replied:
The thing is I am not yet completey sure what the requested index type should be. Is it always just a numeric value? It certainly will be at some point an pair over numeric values. Can it be something more complex? As a matter of fact, we also don't care. We just propagate the index as is to the alignment result. So we don't do any operations on that. The only thing we can require is copyability. That is something I can indeed test.

joergi-w replied:
Okay, that's a good solution.

joergi-w commented on include/seqan3/alignment/pairwise/detail/utility_traits.hpp:
```diff
@@ -0,0 +1,45 @@
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides utility traits for the configuration and execution of the alignment algorithm.
```
The name `utility traits` does not give any insight what is in this file.
Consider renaming the file to chunk_traits.hpp or similar (even if you plan to extend this file later).

rrahn replied:
I renamed it to type_traits. And yes it is already extended in another open PR.

joergi-w commented on include/seqan3/alignment/pairwise/execution/alignment_executor_two_way.hpp:
```diff
@@ -302,11 +300,13 @@ class alignment_executor_two_way
     void move_initialise(alignment_executor_two_way && other) noexcept
     {
         kernel = std::move(other.kernel);
+        _chunk_size = std::move(other._chunk_size);
```
Could you remove the leading underscore from the member variable `chunk_size`? I don't see any reason for this and all other members are also without leading `_`.

rrahn replied:
Actually, I'd love to but there is a function called chunk_size as well and that doesn't work. I do not want to invent a new name for the member and rather stick in these rare cases to what most people do extending the name with an underscore.

rrahn commented:
@joergi-w I addressed all your comments. I have one more question though. Could you next time please use the suggestion feature for small changes like documentation issues? Then I can directly submit them as a commit and don't have to find the places in code that need to be adapted. That would be of great help and speedup applying the changes. Thank you very much.

joergi-w commented:
> @joergi-w I addressed all your comments. I have one more question though. Could you next time please use the suggestion feature for small changes like documentation issues? Then I can directly submit them as a commit and don't have to find the places in code that need to be adapted. That would be of great help and speedup applying the changes. Thank you very much.

I actually would love to use the suggestion tool whenever applicable. Unfortunately, Github doesn't offer it when I display the changes commit-wise... (Error message: You are viewing a subset of changes.) 
Does it work for you or do you know a solution?

rrahn commented:
No not a perfect solution. I would annotate stuff and later make them suggestions. But also agree that this is cumbersome for the reviewer.

smehringer commented:
> I actually would love to use the suggestion tool whenever applicable. Unfortunately, Github doesn't offer it when I display the changes commit-wise... (Error message: You are viewing a subset of changes.)
Does it work for you or do you know a solution?

You can just manually type \`\`\`suggestion ...```!

smehringer commented on include/seqan3/alignment/pairwise/detail/type_traits.hpp:
```diff
@@ -0,0 +1,45 @@
+ *
+ * \details
+ *
+ * This transformation trait transforms a range over sequence pairs into an indexed range over sequence pairs.
```
```suggestion
 * This transformation trait transforms a range over sequence pairs into a range over indexed sequence pairs.
```
?

smehringer commented on include/seqan3/alignment/pairwise/execution/alignment_executor_two_way.hpp:
```diff
@@ -56,19 +57,19 @@ class alignment_executor_two_way
-    using resource_iterator_t = std::ranges::iterator_t<resource_type>;
-    //!\brief The value type of the resource.
-    using resource_value_type = std::ranges::range_value_t<resource_type>;
+    using chunked_chunked_resource_iterator_t = std::ranges::iterator_t<chunked_resource_type>;
```
sometimes you abbreviate with `_t` and sometimes you call it `_type` (`chunked_resource_type` vs `chunked_chunked_resource_iterator_t`)
Maybe change one of them to match the others? (wwithout chenging the whole file but only the lines you touch anyway)

rrahn replied:
right. I kept the naming close to stl. So iterator without any suffix and everything else with _type suffix.

smehringer commented on include/seqan3/alignment/pairwise/execution/alignment_executor_two_way.hpp:
```diff
@@ -257,27 +257,25 @@ class alignment_executor_two_way
-        assert(buffer_limit >= chunk_size());
 
-        while (count < buffer_limit && !is_eof())
+        size_t current_chunk_size = std::ranges::distance(*chunked_resource_it);
```
maybe assert that the `chunked_resource_it` is a forward_iterator? might save some debugging time if we ever use a wrong view on top

smehringer replied:
and you can set this variable inside of the for loop signature.

rrahn replied:
> maybe assert that the chunked_resource_it is a forward_iterator? might save some debugging time if we ever use a wrong view on top

I don't see this as a necessary step as the underlying range is required to be forward range. I don't see why chunk view should ever degrade a forward range to an input range.

rrahn commented on include/seqan3/alignment/pairwise/align_pairwise.hpp:
```diff
@@ -162,36 +164,29 @@ constexpr auto align_pairwise(sequence_t && sequences,
-        {
-            if constexpr (exec_rule::template exists<detail::vectorise_tag>())
-            {
-                return seqan3::par_unseq;
```
Because we don't have a special excecution handler for vectorised code. We simply do the chunking as a high level functionality in the alignment executor. Only the algorithm that receives the chunks does the vectorisation or not.

rrahn commented on include/seqan3/alignment/pairwise/alignment_configurator.hpp:
```diff
@@ -194,13 +218,19 @@ struct alignment_configurator
+     * In case of a missing configuration that has a default, e.g. the seqan3::align_cfg::result option, the
+     * default version of this configuration element is added to the passed configuration object.
+     * The return type is a pair over a std::function object and the configuration object. Thus, the calling function
+     * has access to the possibly modified configuration object. This way it can examine the same configurations as
```
I think the previous sentence was clear enough. I removed the second one.

smehringer commented:
only serparate review commits into the commit structure
